### PR TITLE
Hackathon/test example trimesh utils

### PR DESCRIPTION
**Pull request by zlyfunction**
add util functions in test_example_meshes.cpp
- trimesh_simply_connected_components
- trimesh_n_bd_loops
- trimesh_euler_char
- trimesh_genus
add two test tri_mesh

### CODE REVIEW

File: tests/test_example_meshes.cpp [hunk 1]
```diff
@@ -16,6 +17,121 @@ struct MeshDebugInfo
...
+}
+
+
+int trimesh_simply_connected_components(const DEBUG_TriMesh& m, std::vector<int> &component_ids)
+{
+    component_ids.resize(m.capacity(PrimitiveType::Face), -1);
+    auto all_face_tuples = m.get_all(PrimitiveType::Face);
+    int component_id = 0;
+    // BFS
+    for (size_t i = 0; i < all_face_tuples.size(); ++i) {
```
Comment (mtao) [inline tests/test_example_meshes.cpp:35]: why not use a range for loop?

File: tests/test_example_meshes.cpp [hunk 2]
```diff
@@ -16,6 +17,121 @@ struct MeshDebugInfo
...
+    return component_id;
+}
+
+int trimesh_simply_connected_components(const DEBUG_TriMesh& m)
+{
+    std::vector<int> component_ids;
+    return trimesh_simply_connected_components(m, component_ids);
+}
+
+int trimesh_n_bd_loops(const DEBUG_TriMesh& m)
```
Comment (mtao) [inline tests/test_example_meshes.cpp:76]: please use full words for function names (and typically for variable names as well).
we're not using fortran 77 :)

File: tests/test_example_meshes.cpp [hunk 3]
```diff
@@ -16,6 +17,121 @@ struct MeshDebugInfo
...
+{
+    std::vector<int> bd_loops(m.capacity(PrimitiveType::Edge), -1);
+    auto all_edge_tuples = m.get_all(PrimitiveType::Edge);
+    int bd_loop_id = 0;
+    for (size_t i = 0; i < all_edge_tuples.size(); ++i) 
+    {
+        if (!m.is_boundary(all_edge_tuples[i])) {
```
Comment (mtao) [inline tests/test_example_meshes.cpp:83]: why not range for loop?
also request `all_edge_tuples[i]` multiple times, just create a ref (or just const ref in a range for loop)

Comment (mtao) [reply tests/test_example_meshes.cpp:83]: oh you even declare this as cur_edge as a value a few lines down

File: tests/test_example_meshes.cpp [hunk 4]
```diff
@@ -16,6 +17,121 @@ struct MeshDebugInfo
...
+        }
+        long eid = m.id(all_edge_tuples[i], PrimitiveType::Edge);
+        if (bd_loops[eid] != -1) {
+            continue;
+        } // visited
+
+        Tuple cur_edge = all_edge_tuples[i];
+        long cur_eid = m.id(cur_edge, PrimitiveType::Edge);
+        // find one boundary loop
+        while (bd_loops[m.id(all_edge_tuples[i], PrimitiveType::Edge)] == -1 || cur_eid != m.id(all_edge_tuples[i], PrimitiveType::Edge))
```
Comment (mtao) [inline tests/test_example_meshes.cpp:94]: declare variables when values are used twice.... `m.id(all_edge_tuples[i], PrimitiveType::Edge)` doesn't need to appear more than once

File: tests/test_example_meshes.cpp [hunk 5]
```diff
@@ -16,6 +17,121 @@ struct MeshDebugInfo
...
+            cur_eid = m.id(cur_edge, PrimitiveType::Edge);
+        }
+        
+        bd_loop_id++;
+    }
+
+    return bd_loop_id;
+}
+
+int trimesh_euler_char(const DEBUG_TriMesh& m)
```
Comment (mtao) [inline tests/test_example_meshes.cpp:116]: euler_characteristic...

File: tests/tools/TriMesh_examples.hpp [hunk 6]
```diff
@@ -88,4 +88,11 @@ TriMesh hex_plus_two();
 //   \ / \ / \ /
 //    7---8---9
 TriMesh edge_region();
+
+//  0---2  3---4
+//  | /    | \ |
+//  1      5---6
+TriMesh three_triangles_with_two_components();
+
```
Comment (mtao) [inline tests/tools/TriMesh_examples.hpp:96]: what's the topology for nine with a hole? would be nice to be consistent about having figures for everything

**Review by mtao — APPROVED** [review]
minor changes to naming/style required but generally good  to go

File: tests/test_example_meshes.cpp [hunk 7]
```diff
@@ -16,6 +17,121 @@ struct MeshDebugInfo
...
+            {
+                if (!m.is_boundary(f_tuple))
+                {   
+                    auto adj_face_tuple = m.switch_tuple(f_tuple, PrimitiveType::Face);
+                    long adj_fid = m.id(adj_face_tuple, PrimitiveType::Face);
+                    if (component_ids[adj_fid] == -1) {
+                        q.push(adj_fid);
+                    }  
+                }
+                f_tuple = m.switch_tuple(m.switch_tuple(f_tuple, PrimitiveType::Vertex), PrimitiveType::Edge);
```
Comment (daniel-zint) [inline tests/test_example_meshes.cpp:60]: use `m.switch_vertex` and `m.switch_edge` for better readability

File: tests/test_example_meshes.cpp [hunk 8]
```diff
@@ -16,6 +17,121 @@ struct MeshDebugInfo
...
+            } 
+        }
+        
+        component_id++;
```
Comment (daniel-zint) [inline tests/test_example_meshes.cpp:64]: `++component_id`

File: tests/test_example_meshes.cpp (hunk 2 above)
Comment (daniel-zint) [reply tests/test_example_meshes.cpp:76]: Exception: you have a lot of variables and using acronyms increases readability. In that case, add explanations to the variable definitions. But this is only rarely the case and usually already a sign that the code needs to be refactored.

File: tests/test_example_meshes.cpp [hunk 9]
```diff
@@ -16,6 +17,121 @@ struct MeshDebugInfo
...
+    int n_bd_loops = trimesh_n_bd_loops(m);
+    return simplex_counts[0] - simplex_counts[1] + simplex_counts[2] + n_bd_loops;
+}
+
+int trimesh_genus(const DEBUG_TriMesh& m)
+{
+    if (trimesh_simply_connected_components(m) != 1)
+    {
+        wmtk::logger().error("not a connected surface\n");
+        throw std::runtime_error("GenusComputeError");
```
Comment (daniel-zint) [inline tests/test_example_meshes.cpp:128]: A minor thing but logging and throwing an error seems to me a bit overkill. I feel like throwing is enough. Anyways, the messages of the two should be the same I think.

File: tests/test_example_meshes.cpp [hunk 10]
```diff
@@ -16,6 +17,121 @@ struct MeshDebugInfo
...
+    }
+    return (2 - trimesh_euler_char(m)) / 2;
+}
+
+
+
 void run_debug_trimesh(const DEBUG_TriMesh& m, const MeshDebugInfo& info)
```
Comment (daniel-zint) [inline tests/test_example_meshes.cpp:135]: give it a more explanatory name or add a comment

File: tests/test_example_meshes.cpp [hunk 11]
```diff
@@ -80,9 +195,32 @@ TEST_CASE("test_debug_trimeshes_one_ear")
...
-    info.simplex_counts = std::array<long, 3>{{3, 3, 3}};
+    info.simplex_counts = std::array<long, 3>{{4, 5, 2}};
     run_debug_trimesh(m, info);
 }
+
+TEST_CASE("test_debug_trimeshes_two_components")
+{
+    DEBUG_TriMesh m;
+    m = three_triangles_with_two_components();
+    MeshDebugInfo info;
```
Comment (daniel-zint) [inline tests/test_example_meshes.cpp:210]: `info` is not used in that test

File: tests/tools/TriMesh_examples.hpp (hunk 6 above)
Comment (daniel-zint) [reply tests/tools/TriMesh_examples.hpp:96]: \+

**Review by daniel-zint — COMMENTED** [review]
some more comments, minor things though

File: tests/test_example_meshes.cpp (hunk 2 above)
Comment (mtao) [reply tests/test_example_meshes.cpp:76]: sure, exception to the exception: only for short functions :)

File: tests/test_example_meshes.cpp [hunk 12]
```diff
@@ -16,6 +17,121 @@ struct MeshDebugInfo
...
+                    }  
+                }
+                f_tuple = m.switch_tuple(m.switch_tuple(f_tuple, PrimitiveType::Vertex), PrimitiveType::Edge);
+            } 
+        }
+        
+        component_id++;
+    }
+
+    return component_id;
```
Comment (mtao) [inline tests/test_example_meshes.cpp:67]: actually, why are we returning the number of components AND the component id maps. just return the component ids (and avoid this return type parameter mess). size() suffices

File: tests/test_example_meshes.cpp (hunk 9 above)
Comment (mtao) [reply tests/test_example_meshes.cpp:128]: i agree but only if we have a top level try/catch that forwards exceptions to the logger.
there are cases where you want to send someone an output log (maybe a log from a file) and that won't contain the exceptions)..

on the other hand, the runtime error message is not practical (you're declaring it like you're specifying the runtime error is a class name, even though the aforementioned log message spits out more information. might as well make these two statements the same

**Comment by mtao** [comment]
BTW the computation of topological invariants might be helpful outside fof unit etsts so it might be nice to leave them as utils sort of functions. It might also be nice to write a more efficient implementation of simplex counts at some point reading off flags, but this is a fairly low priority.

File: tests/test_example_meshes.cpp (hunk 9 above)
Comment (daniel-zint) [reply tests/test_example_meshes.cpp:128]: True, I didn't consider that. Then I guess having the logger and the error simultaneously with the same message might be the best option.

File: tests/test_example_meshes.cpp (hunk 12 above)
Comment (zlyfunction) [reply tests/test_example_meshes.cpp:67]: components maps is a per face attributes, to get the return size the right way is to map.max()+1, which cost extra time?

**Comment by zlyfunction** [comment]
updated based on comments

File: tests/tools/TriMesh_examples.hpp (hunk 6 above)
Comment (mtao) [reply tests/tools/TriMesh_examples.hpp:96]: Uh, could we have a simpler picture? I think Daniel had a new test with a figure that could be cut down for this purpose

**Review by mtao — COMMENTED** [review]
Responses to a few updates

File: tests/test_example_meshes.cpp (hunk 9 above)
Comment (mtao) [reply tests/test_example_meshes.cpp:133]: alternative is to have a global catch that rethrows? i'm not sure how this works with debuggers though

**Pull request merged**